cli/lsp/diagnostics.rs

```diff
@@ -313,9 +374,14 @@ impl DiagnosticsServer {
       performance,
       ts_server,
       batch_counter: Default::default(),
+      state: Default::default(),
     }
   }
 
+  pub fn state(&self) -> Arc<RwLock<DiagnosticsState>> {
```

Comment:
Exposing this RwLock is very sketchy because it increases the chance of deadlocks. It would be good if we could bury this behind an API somehow.

Reply:
I opened https://github.com/denoland/deno/issues/20658

Reply:
I can replace this with `DiagnosticsState`'s reading methods, callers shouldn't be able to write it anyway. I'll do it in the patch for https://github.com/denoland/deno/pull/20632#issuecomment-1732208452.

Reply:
What's the difference between `diagnostics_by_specifier` and `DiagnosticsState`?

Reply:
> What's the difference between `diagnostics_by_specifier` and `DiagnosticsState`?

`DiagnosticsPublisher` is a transient structure along with `diagnostics_by_specifier`, it copies stuff into persistent `state` on publish.

Reply:
I think it used to be transient? I changed it to be persistent to removing flickering.

It seems like we could combine the two?